// qminder-api/build/qminder-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = exports.setServer = exports.setKey = exports.graphql = exports.webhooks = exports.users = exports.tickets = exports.locations = exports.lines = exports.devices = exports.ConnectionStatus = exports.Webhook = exports.User = exports.Ticket = exports.Location = exports.Line = exports.Device = exports.Desk = exports.ApiBase = void 0;
// Import data structures
const Desk_1 = require("./model/Desk");
exports.Desk = Desk_1.default;
const Device_1 = require("./model/Device");
exports.Device = Device_1.default;
const Line_1 = require("./model/Line");
exports.Line = Line_1.default;
const Location_1 = require("./model/Location");
exports.Location = Location_1.default;
const Ticket_1 = require("./model/Ticket");
exports.Ticket = Ticket_1.default;
const User_1 = require("./model/User");
exports.User = User_1.default;
const Webhook_1 = require("./model/Webhook");
exports.Webhook = Webhook_1.default;
// Import services
const api_base_1 = require("./api-base");
exports.ApiBase = api_base_1.default;
const DeviceService_1 = require("./services/DeviceService");
exports.devices = DeviceService_1.default;
const LineService_1 = require("./services/LineService");
exports.lines = LineService_1.default;
const LocationService_1 = require("./services/LocationService");
exports.locations = LocationService_1.default;
const TicketService_1 = require("./services/TicketService");
exports.tickets = TicketService_1.default;
const UserService_1 = require("./services/UserService");
exports.users = UserService_1.default;
const WebhooksService_1 = require("./services/WebhooksService");
exports.webhooks = WebhooksService_1.default;
const GraphQLService_1 = require("./services/GraphQLService");
exports.graphql = GraphQLService_1.default;
Object.defineProperty(exports, "ConnectionStatus", { enumerable: true, get: function () { return GraphQLService_1.ConnectionStatus; } });
/**
 * Set the Qminder API key.
 *
 * This function sets the API key and enables you to use the API methods.
 *
 * @param key  the Qminder API key
 */
const setKey = (key) => {
    api_base_1.default.setKey(key);
    GraphQLService_1.default.setKey(key);
};
exports.setKey = setKey;
/**
 * Set the Qminder API server hostname.
 *
 * This is used to point the API at a different server.
 *
 * @param server  the API server hostname, for example "api.qminder.com"
 * @hidden
 */
const setServer = (server) => {
    api_base_1.default.setServer(server);
    GraphQLService_1.default.setServer(`wss://${server}:443`);
};
exports.setServer = setServer;
/** @hidden */
exports.version = '4.0.1';
